src/sparse_blas/backends/mkl_common/mkl_spmv.cxx

```diff
@@ -0,0 +1,217 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+// The operation descriptor is not needed as long as the backend does not have an equivalent type and does not support external workspace.
+using spmv_descr = void *;
+
+void init_spmv_descr(sycl::queue & /*queue*/, oneapi::mkl::sparse::spmv_descr_t *p_spmv_descr) {
+    *p_spmv_descr = nullptr;
+}
+
+sycl::event release_spmv_descr(sycl::queue &queue, oneapi::mkl::sparse::spmv_descr_t /*spmv_descr*/,
+                               const std::vector<sycl::event> &dependencies) {
+    return detail::collapse_dependencies(queue, dependencies);
+}
+
+void check_valid_spmv(const std::string function_name, sycl::queue &queue,
+                      oneapi::mkl::transpose opA, oneapi::mkl::sparse::matrix_view A_view,
+                      oneapi::mkl::sparse::matrix_handle_t A_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+                      oneapi::mkl::sparse::dense_vector_handle_t y_handle, const void *alpha,
+                      const void *beta) {
+    THROW_IF_NULLPTR(function_name, A_handle);
+    THROW_IF_NULLPTR(function_name, x_handle);
+    THROW_IF_NULLPTR(function_name, y_handle);
+
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    detail::check_all_containers_compatible(function_name, internal_A_handle, x_handle, y_handle);
+    if (internal_A_handle->all_use_buffer()) {
+        detail::check_ptr_is_host_accessible("spmv", "alpha", queue, alpha);
+        detail::check_ptr_is_host_accessible("spmv", "beta", queue, beta);
+    }
+    if (detail::is_ptr_accessible_on_host(queue, alpha) !=
+        detail::is_ptr_accessible_on_host(queue, beta)) {
+        throw mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "Alpha and beta must both be placed on host memory or device memory.");
+    }
```

Comment:
Have we documented this in the oneAPI specification that `alpha`, `beta` must both be on device or both be on host? I don't think so, see https://uxlfoundation.github.io/oneAPI-spec/spec/elements/oneMKL/source/domains/spblas/operations/spmv.html#onemkl-sparse-spmv-header. Either we add that restriction into the oneAPI specofication (what do cuSPARSE/rocSPARSE say?) or we add support for one being on host and one on device here.

Reply:
The cuSPARSE and rocSPARSE backends cannot use different memory kinds hence this restriction. See more details in this spec PR: https://github.com/uxlfoundation/oneAPI-spec/pull/556

Reply:
Resoving old comment